### PR TITLE
New: Implement cacheStrategy (refs eslint/rfcs#63)

### PR DESCRIPTION
**Pull request by chambo-e**
Hello 👋 

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[X] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This is a follow up of #13713 which seems abandoned.

I corrected the tests by binding every stubs to the newly created sinon sandbox

### CODE REVIEW

File: docs/user-guide/command-line-interface.md
Comment (nzakas) [inline docs/user-guide/command-line-interface.md:450]:
```suggestion
Example:
```

File: lib/cli-engine/lint-result-cache.js
Comment (nzakas) [inline lib/cli-engine/lint-result-cache.js:37]: Because `configHelper` is an object rather than a string, perhaps `hasValidCacheStrategy()` would be a better name for this function?

Another option would be to just pass in the cache strategy string rather than the entire configHelper object.

**Review by nzakas — CHANGES_REQUESTED** [review]
Overall looks good, just a couple of things to clarify.

File: lib/cli-engine/lint-result-cache.js
Comment (chambo-e) [reply lib/cli-engine/lint-result-cache.js:37]: You're right, I'll opt for the second option, easier to understand/reuse :)

**Review by royriojas — APPROVED** [review]
Very nice

**Review by nzakas — APPROVED** [review]
LGTM. I'd like another set of eyes on this before merging.

File: lib/eslint/eslint.js
Comment (mdjermanovic) [inline lib/eslint/eslint.js:161]: We should add `cacheStrategy` to `return { ... }`.

Otherwise, it has no effect as it won't be passed to `CLIEngine` [here](https://github.com/eslint/eslint/blob/1d11d48a41f02fed627a37072f1fa1d7bd0dca2d/lib/eslint/eslint.js#L431). I tested `--cache-strategy content` on CLI and it still stores `mtime` and `size` instead of `hash` in the cache file.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
It would be good to also fix the behavior in the case that is described here:

https://github.com/eslint/rfcs/pull/63#issuecomment-676463622

**Comment by nzakas** [comment]
@mdjermanovic you indicated in the RFC that this wasn't a blocker. Is that something we can manage via a separate issue rather than blocking this PR for it?

**Comment by mdjermanovic** [comment]
Agreed we shouldn't address that in this PR. The issue might get fixed in `file-entry-cache` (https://github.com/royriojas/file-entry-cache/pull/22).

I meant it shouldn't be a blocker for this feature. If it isn't fixable, we can tell users to always delete the cache file before changing the strategy.

**Comment by royriojas** [comment]
@mdjermanovic @nzakas published `file-entry-cache@6.0.1` with @mdjermanovic fix

**Comment by mdjermanovic** [comment]
@royriojas thanks!

@chambo-e can you update `file-entry-cache` to `^6.0.1` in [package.json](https://github.com/eslint/eslint/blob/87c43a5d7ea2018cffd6d9b5c431ecb60caaf0d6/package.json#L63)?

File: lib/eslint/eslint.js
Comment (chambo-e) [reply lib/eslint/eslint.js:161]: You're absolutely right, this has been fixed :)

File: lib/cli-engine/lint-result-cache.js
Comment (chambo-e) [reply lib/cli-engine/lint-result-cache.js]: I like that, simpler :)

**Comment by chambo-e** [comment]
@mdjermanovic sorry about the delay, I've corrected your feedbacks :)

File: lib/eslint/eslint.js
Comment (mdjermanovic) [reply lib/eslint/eslint.js:46]: Thanks for updating `ESLint` constructor docs!

I think we can discuss `CLIEngine` docs in a follow-up.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
A few suggestions, and then LGTM.

**Comment by chambo-e** [comment]
@mdjermanovic all done

File: docs/user-guide/command-line-interface.md
Comment (bmish) [inline docs/user-guide/command-line-interface.md:448]: I wonder if it would be helpful to mention the common reason why this happens, which I believe is this:

> 
```suggestion
The `content` strategy can be useful in cases where the modification time of your files change even if their contents have not. For example, this can happen during git operations like git clone because git does not track file modification time.
```

Comment (nzakas) [reply docs/user-guide/command-line-interface.md:448]: I like it! :+1:

**Review by nzakas — APPROVED** [review]

File: docs/user-guide/command-line-interface.md
Comment (chambo-e) [reply docs/user-guide/command-line-interface.md]: Thanks 😃

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Review by snitin315 — APPROVED** [review]
💯

**Comment by mdjermanovic** [comment]
Thanks for contributing!

**Pull request merged**
